### PR TITLE
Check enumerateDevices behavior pre-capture, cross-origin and post site data clearing

### PR DESCRIPTION
**Pull request by dontcallmedom**

### CODE REVIEW

File: mediacapture-streams/MediaDevices-enumerateDevices-returned-objects.https.html [hunk 1]
```diff
@@ -13,8 +13,10 @@
 function doTest(callGetUserMedia, testName)
 {
     promise_test(async () => {
-        if (callGetUserMedia) {
-            await setMediaPermission();
+        if (!callGetUserMedia) {
+            await setMediaPermission("denied");
```
Comment (jan-ivar) [inline mediacapture-streams/MediaDevices-enumerateDevices-returned-objects.https.html:17]: This test is testing the wrong thing / is too permissive.

enumerateDevices exposure is not tied to permission, but to [device-information-can-be-exposed](https://w3c.github.io/mediacapture-main/getusermedia.html#device-information-can-be-exposed), i.e. [[[canExposeCameraInfo](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-canexposecamerainfo)]] OR [[[canExposeMicrophoneInfo](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-canexposemicrophoneinfo)]] OR live-checks.

Specifically, we need to test that information is not exposed in cases where persistent permission is granted, but getUserMedia has not been called from this document before. See steps in [crbug 1101860](https://crbug.com/1101860).

File: mediacapture-streams/MediaDevices-enumerateDevices-returned-objects.https.html [hunk 2]
```diff
@@ -28,12 +30,31 @@
...
+              assert_equals(deviceKinds.filter(x => x===kind).length, 1, "At most 1 " + kind + " prior to capture");
+            }
         }
     }, testName);
 }
 
-doTest(false, "enumerateDevices returns new objects in case device-info permission is not granted");
-doTest(true, "enumerateDevices returns new objects in case device-info permission is granted");
+doTest(false, "enumerateDevices returns expected mostly empty objects in case device-info permission is not granted");
+doTest(true, "enumerateDevices returns expected objects in case device-info permission is granted");
```
Comment (jan-ivar) [inline mediacapture-streams/MediaDevices-enumerateDevices-returned-objects.https.html:57]: I think this would be easier to read if it were flattened into two independent tests.

**Review by jan-ivar — CHANGES_REQUESTED** [review]
Thanks for writing these! The cookie parts look good, but the tests on enumerateDevices information exposure need to be decoupled from permission.

File: mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+<script src=/resources/testdriver-vendor.js></script>
+<script src=permission-helper.js></script>
+</head>
+<body>
+  <iframe allow="camera 'src';microphone 'src'" id=same src="/mediacapture-streams/iframe-enumerate.html"></iframe>
+<iframe allow="camera 'src';microphone 'src'" id=cross src="https://{{hosts[][www1]}}:{{ports[https][0]}}/mediacapture-streams/iframe-enumerate.html"></iframe>
+<script>
+  function compareDeviceList(list1, list2, origintype) {
+    assert_equals(list1.length, list2.length, "Same number of devices detected in " + origintype + "-origin");
+    for (let device of list1) {
```
[the remark this answers is not in the document]
Comment (dontcallmedom) [reply mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html]: done as part of de8e78e

File: mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+<script src=/resources/testharnessreport.js></script>
+<script src=/resources/testdriver.js></script>
+<script src=/resources/testdriver-vendor.js></script>
+<script src=permission-helper.js></script>
+</head>
+<body>
+  <iframe allow="camera 'src';microphone 'src'" id=same src="/mediacapture-streams/iframe-enumerate.html"></iframe>
+<iframe allow="camera 'src';microphone 'src'" id=cross src="https://{{hosts[][www1]}}:{{ports[https][0]}}/mediacapture-streams/iframe-enumerate.html"></iframe>
+<script>
+  function compareDeviceList(list1, list2, origintype) {
```
[the remark this answers is not in the document]
Comment (dontcallmedom) [reply mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html]: done as part of de8e78e

File: mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    assert_equals(list1.length, list2.length, "Same number of devices detected in " + origintype + "-origin");
+    for (let device of list1) {
+      const sameDevice = list2.find(d => d.deviceId === device.deviceId);
+      assert_true(!!list2.find(d => d.label === device.label), "labels stay the same across origins");
+      // TODO: is using "default" compliant with the spec?
+      if (device.deviceId !== "default") {
+        if (origintype === "same") {
+          assert_true(!!sameDevice, "deviceIds stay the same when loaded in same origin");
+          assert_equals(sameDevice.label, device.label, "labels stay the same when loaded in same origin");
+          assert_equals(sameDevice.kind, device.kind, "kind stay the same when loaded in same origin");
```
[the remark this answers is not in the document]
Comment (dontcallmedom) [reply mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html]: done as part of de8e78e

File: mediacapture-streams/MediaDevices-enumerateDevices-returned-objects.https.html (hunk 2 above)
Comment (dontcallmedom) [reply mediacapture-streams/MediaDevices-enumerateDevices-returned-objects.https.html:56]: I don't disagree, but since the test started that way, I don't feel the urgent need to flatten it :)

File: mediacapture-streams/MediaDevices-enumerateDevices-returned-objects.https.html (hunk 1 above)
Comment (dontcallmedom) [reply mediacapture-streams/MediaDevices-enumerateDevices-returned-objects.https.html]: I've addressed the permission-coupling in 043b8f1 and proposed an additional test in dce98c1  to cater for the persistent permission situation - PTAL

File: mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+          assert_not_equals(sameDevice.groupId, device.groupId, "groupId is specific to a document");
+          // The group identifier MUST be uniquely generated for each document.
+        } else {
+          assert_false(!!sameDevice, "deviceIds are not " + (origintype === "cross" ? "shared across origin" : "kept after clearing site data"));
```
[the remark this answers is not in the document]
Comment (dontcallmedom) [reply mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html]: done (I think?) as part of de8e78e

File: mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html [hunk 7]
```diff
@@ -0,0 +1,78 @@
+<!doctype html>
+<html>
+<head>
+<title>enumerateDevices is returning new MediaDeviceInfo objects every time</title>
```
[the remark this answers is not in the document]
Comment (dontcallmedom) [reply mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html]: done in b3ae109

File: mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+  <iframe allow="camera 'src';microphone 'src'" id=same src="/mediacapture-streams/iframe-enumerate.html"></iframe>
+<iframe allow="camera 'src';microphone 'src'" id=cross src="https://{{hosts[][www1]}}:{{ports[https][0]}}/mediacapture-streams/iframe-enumerate.html"></iframe>
+<script>
+  function compareDeviceList(list1, list2, origintype) {
+    assert_equals(list1.length, list2.length, "Same number of devices detected in " + origintype + "-origin");
+    for (let device of list1) {
+      const sameDevice = list2.find(d => d.deviceId === device.deviceId);
+      assert_true(!!list2.find(d => d.label === device.label), "labels stay the same across origins");
+      // TODO: is using "default" compliant with the spec?
+      if (device.deviceId !== "default") {
```
[the remark this answers is not in the document]
Comment (dontcallmedom) [reply mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html]: I've removed that check for now  in de8e78e but it's not that clear to me that it is not compliant - that is, the spec clearly doesn't mandate or suggest having a `deviceId` of that value, but [it seems to allow](https://w3c.github.io/mediacapture-main/#dom-mediadeviceinfo-deviceid) "magic" `deviceId`s with:
> An identifier can be reused across origins as long as it is not tied to the user and can be guessed by other means, like the User-Agent string

so I'm not sure my test that `deviceId`s should be strictly different across origins / after cookie clearing works without considering these exceptions; I'm not sure how to manage that exception without hardcoding the known magic values.

File: mediacapture-streams/MediaDevices-enumerateDevices-persistent-permission.https.html [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    const loadWatcher = new EventWatcher(t, iframe, ['load']);
+    await loadWatcher.wait_for('load');
+    const msgWatcher = new EventWatcher(t, window, ['message']);
+    frames[0].postMessage('run', '*')
+    const e = await msgWatcher.wait_for('message');
+    const iframeDevices = e.data.devices;
+    const deviceKinds = iframeDevices.map(d => d.kind);
+    for (let kind of deviceKinds) {
+      assert_equals(deviceKinds.filter(x => x===kind).length, 1, "At most 1 " + kind + " prior to capture");
+    }
```
Comment (jan-ivar) [inline mediacapture-streams/MediaDevices-enumerateDevices-persistent-permission.https.html:33]: This triggers two asserts per failure. How about:
```suggestion
    const kinds = iframeDevices.map(({kind}) => kind);
    assert_equals(kinds.length == new Set(kinds).size, "At most one of a kind prior to capture");
```

File: mediacapture-streams/MediaDevices-enumerateDevices-returned-objects.https.html [hunk 10]
```diff
@@ -14,7 +14,7 @@
 {
     promise_test(async () => {
         if (callGetUserMedia) {
-            await setMediaPermission();
+            await setMediaPermission("granted");
```
Comment (jan-ivar) [inline mediacapture-streams/MediaDevices-enumerateDevices-returned-objects.https.html:17]: Why this change? Isn't `"granted"` already the default? Whichever way we go we should be consistent.

File: mediacapture-streams/MediaDevices-enumerateDevices-returned-objects.https.html [hunk 11]
```diff
@@ -14,7 +14,7 @@
...
             await navigator.mediaDevices.getUserMedia({audio : true, video: true});
```
Comment (jan-ivar) [inline mediacapture-streams/MediaDevices-enumerateDevices-returned-objects.https.html:18]: This test might be racy and fail wpt stability mode, as mentioned above. We should hold on to the returned stream and call stop on its tracks. This might be simpler to do if we flatten the test first.

File: mediacapture-streams/message-enumerateddevices.js [hunk 12]
```diff
@@ -0,0 +1,7 @@
+onmessage = async e => {
+  await navigator.mediaDevices.getUserMedia({audio: true, video: true});
+  const devices = await navigator.mediaDevices.enumerateDevices();
```
Comment (jan-ivar) [inline mediacapture-streams/message-enumerateddevices.js:3]: I'd like to see us hold the stream alive across the enumerateDevices call here for correctness:
```suggestion
  const stream = await navigator.mediaDevices.getUserMedia({audio: true, video: true});
  const devices = await navigator.mediaDevices.enumerateDevices();
  stream.getTracks().forEach(t => t.stop());
```

**Review by jan-ivar — CHANGES_REQUESTED** [review]
Thanks for the additional test!

Found some races. Sorry for not catching these on my first review.

File: mediacapture-streams/MediaDevices-enumerateDevices-returned-objects.https.html (hunk 11 above)
Comment (jan-ivar) [reply mediacapture-streams/MediaDevices-enumerateDevices-returned-objects.https.html:18]: This looks like an existing issue, so I'd take a follow-up on this if you prefer.

File: mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html (hunk 8 above)
[the remark this answers is not in the document]
Comment (jan-ivar) [reply mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html]: I forgot we added that. Ugh. I'll accept adding it back in. Beats allowing more of them.

File: mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+      assert_true(!!list2.find(d => d.label === device.label), "labels stay the same across origins");
+    }
+  }
+
+  let deviceList;
+
+  promise_test(async t => {
+    await setMediaPermission();
+    await navigator.mediaDevices.getUserMedia({audio : true, video: true});
+    deviceList =  await navigator.mediaDevices.enumerateDevices();
```
[the remark this answers is not in the document]
Comment (jan-ivar) [reply mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html:27]: I may be wrong here depending on how wpt stability runs loop tests. If they create a new document for each run then there's no issue, but if they don't then these tests aren't deterministic if run in a tight loop, unless we stop tracks in between runs.

**Review by jan-ivar — APPROVED** [review]

File: mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html (hunk 8 above)
[the remark this answers is not in the document]
Comment (dontcallmedom) [reply mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html]: re-instated in f18c67aa3a

File: mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+<script src=permission-helper.js></script>
+</head>
+<body>
+  <iframe allow="camera 'src';microphone 'src'" id=same src="/mediacapture-streams/iframe-enumerate.html"></iframe>
+<iframe allow="camera 'src';microphone 'src'" id=cross src="https://{{hosts[][www1]}}:{{ports[https][0]}}/mediacapture-streams/iframe-enumerate.html"></iframe>
+<script>
+  function checkDeviceListsMatch(list1, list2, origintype) {
+    assert_equals(list1.length, list2.length, "Same number of devices detected in " + origintype + "-origin");
+    for (const device of list1) {
+      assert_true(!!list2.find(d => d.label === device.label), "labels stay the same across origins");
```
[the remark this answers is not in the document]
Comment (dontcallmedom) [reply mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html]: removed in bd26c4f6dc

File: mediacapture-streams/MediaDevices-enumerateDevices-returned-objects.https.html (hunk 10 above)
Comment (dontcallmedom) [reply mediacapture-streams/MediaDevices-enumerateDevices-returned-objects.https.html]: reverted in de937849ae
